### DISCUSSION TITLE
can cartographer work with airsim?

### DISCUSSION THREAD

**Discussion by jaredsjohansen** [microsoft/AirSim#3857] [Support Q&A]
## Question
Does airsim produce the correct tf tree to work with ROS cartographer?  If not, is there some setting I can tweak to modify this?

### Include context on what you are trying to achieve
I am running Airsim with a single drone (which has a lidar and imu).
I am also running [airsim-to-ros]( https://microsoft.github.io/AirSim/airsim_ros_pkgs.html).
I want to hook up the output of airsim to cartographer and let cartographer build a map of the environment.
With only these things running, I get the following tf tree: [frames.pdf](https://github.com/microsoft/AirSim/files/6690525/frames.pdf).

#### Context details
I am running on Ubuntu, ROS Melodic, and the latest stable cartographer build. 
I understand that I want my tf tree to look like this: map -> odom -> base_link.

SETTINGS: My settings.json file looks like this:
```{
  "SeeDocsAt": "https://github.com/Microsoft/AirSim/blob/master/docs/settings.md",
  "SettingsVersion": 1.2,
  "SimMode": "Multirotor",
  "ViewMode": "SpringArmChase",
  "ClockSpeed": 1.0,
  "Vehicles": {
    "drone1": {
      "VehicleType": "SimpleFlight",
      "DefaultVehicleState": "Armed",
      "EnableCollisionPassthrogh": false,
      "EnableCollisions": true,
      "AllowAPIAlways": true,
      "RC": {
        "RemoteControlID": 0,
        "AllowAPIWhenDisconnected": false
      },
      "Sensors": {
        "Imu" : {
          "SensorType": 2,
          "Enabled": true
        },
        "Lidar": {
          "SensorType": 6,
          "Enabled": true,
          "NumberOfChannels": 3,
	  "Range": 12,
          "PointsPerSecond": 10000,
          "X": 0,
          "Y": 0,
          "Z": -1,
          "DrawDebugPoints": true
        }
      },
      "X": 2, "Y": 0, "Z": 0,
      "Pitch": 0, "Roll": 0, "Yaw": 0
    }
  },
  "SubWindows": [
    {"WindowID": 0, "ImageType": 0, "CameraName": "front_left_custom", "Visible": true},
    {"WindowID": 1, "ImageType": 0, "CameraName": "front_center_custom", "Visible": false},
    {"WindowID": 2, "ImageType": 0, "CameraName": "front_right_custom", "Visible": true}
  ]
}
```

### Include details of what you already did to find answers
I can't quite determine whether airsim is producing a tf tree that breaks the map->odom->base_link paradigm...or if I've messed something with my cartographer setup...since cartographer is not working as I expect. 

There is a [thread](https://www.scrimmagesim.org/sphinx/html/tutorials/airsim-plugin.html) that says "Set ros_cartographer="true" for ROS AirSim to organize it’s Transform Tree (TF2) structure in a way that is compatible with ROS Cartographer."  That seems to imply to me that airsim may not produce the tf tree (out of the box) that is compatible with cartographer.  If this is true, how can I configure airsim to produce a tf tree that is compatible with cartographer?

I've read the [Cartographer docs](https://google-cartographer-ros.readthedocs.io/en/latest/configuration.html) and tried different things, but can't quite determine whether I have configured cartographer incorrectly...or whether there is something incompatible with the tf tree produced by airsim.

As a reference for what I've done in cartographer, I include the following details below:

URDF: Because there are already tf relationships between drone1, odom_local_ned, and Lidar, my urdf file is empty.  (I believe I don't need to define them.  Please correct me if I am wrong.)

LUA: I have created a my_robot.lua file that looks like this: 
```options = {
  map_builder = MAP_BUILDER,
  trajectory_builder = TRAJECTORY_BUILDER,
  map_frame = "map",
  tracking_frame = "drone1",
  published_frame = "odom_local_ned",
  odom_frame = "not-used",
  provide_odom_frame = false,
  publish_frame_projected_to_2d = false,
  use_pose_extrapolator = true,
  use_odometry = true,
  use_nav_sat = false,
  use_landmarks = false,
  num_laser_scans = 0,
  num_multi_echo_laser_scans = 0,
  num_subdivisions_per_laser_scan = 1,
  num_point_clouds = 1,
  lookup_transform_timeout_sec = 0.2,
  submap_publish_period_sec = 0.3,
  pose_publish_period_sec = 5e-3,
  trajectory_publish_period_sec = 30e-3,
  rangefinder_sampling_ratio = 1.,
  odometry_sampling_ratio = 1.,
  fixed_frame_pose_sampling_ratio = 1.,
  imu_sampling_ratio = 1.,
  landmarks_sampling_ratio = 1.,
}
```

When I run cartographer, I get this error.
`[ WARN] [1624326678.959343053]: W0621 18:51:18.000000 71181 tf_bridge.cc:52] "odom_local_ned" passed to lookupTransform argument source_frame does not exist. `

**Comment by LucasMa00** [comment]
Hello,  I also have tf conflict when run cartographer with airsim. Have you solve this problem? I want to know how to edit the settings to solve this problem.
Thanks in advance!

**Reply by jaredsjohansen** [reply]
There are a few hacks you can do:
(1) You can redirect airsim to publish `tf` data on something different like `tf_garbage`.  You can then write a rosnode that subscribes to `tf_garbage` and publishes only the portions of the tf tree that you care about.  This same rosnode can publish any other portion of the tf tree that you care about.
(2) You may find that the `frame_id` of the sensor data isn't correct (i.e., it doesn't fit into the tf tree appropriately).  As a workaround, we did something similar to (1): we created a rosnode that subscribed to the sensor data, changed the frame_id, reformatted it (if necessary), and published it on a new topic (e.g., subscribed to `odom`; publish updated data on `my_new_odom`).  For any downstream rosnodes that relied on `odom`, we had to add a <remap> from `odom` to `my_new_odom`.

This was very hacky and didn't give us a very clean implementation, but it worked for our short-term project.
